jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/message/MessageInputStream.java

```diff
@@ -209,6 +223,13 @@ public int read(byte[] b, int off, int len) throws IOException
             // grab a fresh buffer
             while (activeBuffer == null || !activeBuffer.hasRemaining())
             {
+                if (activeBuffer != null)
+                {
+                    // All content consumed, release back to pool.
+                    bufferPool.release(activeBuffer);
```

Comment:
why are we only releasing the active buffer before it is drained? Isn't there a case where we return below after consuming the active buffer and if read is not called again then it will not be released until close?

Isn't it better to release after the  ` if (!activeBuffer.hasRemaining())` branch below?

Reply:
The loop condition here `while (activeBuffer == null || !activeBuffer.hasRemaining())` implies that we could have a non null `activeBuffer` in which case we would need to release it before getting a new one from the queue.

This whole loop seems strange to me, if the suspend/resume mechanism is working properly we should never have more than 1 buffer in the Queue at a time, and so we should be able to remove the queue and only use `activeBuffer`.

Reply:
That sounds even better.  Can you verify that is the case? ie that you never have an activeBuffer and a non-zero queue?

Reply:
Note that if you released active buffers when they are consumed, then there would be no need to release when you loop, as it will already have been released.

Reply:
I see why we need the Queue now, we put EOF buffer in with the last frame so we're not waiting forever.

But I have made the change to release buffers after they are consumed.

Reply:
Wow, that is a bit of a cost to just deal with EOF state!  Couldn't that be another field that you only look at once activeBuffer is consumed.

Also, this still looks like the wrong spot for the release. Why is it not done at line 267 below?